### PR TITLE
fix: server mode not shutting down when parent process exits

### PR DESCRIPTION
**Pull request by philasmar**
*Issue #, if available:*
IDE-4723

*Description of changes:*
fix: server mode not shutting down when parent process exits

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs [hunk 1]
```diff
@@ -136,7 +136,9 @@ private async Task<bool> WaitOnParentPid(CancellationToken token)
             {
                 try
                 {
-                    Process.GetProcessById((int)_parentPid);
+                    var process = Process.GetProcessById((int)_parentPid);
+                    if (process.HasExited)
+                        return true;
                     await Task.Delay(1000, token);
                 }
                 catch
```
Comment (ganeshnj) [inline src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs:144]: Can you remove the infinite while and use `WaitForExit`?

https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.waitforexit?view=net-5.0#system-diagnostics-process-waitforexit(system-int32)

**Review by ganeshnj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs (hunk 1 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs]: WaitForExit is a blocking call. I reworked it to use events to remove the while loop.

File: src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs [hunk 2]
```diff
@@ -59,33 +59,29 @@ public async Task ExecuteAsync(CancellationToken cancellationToken = default(Can
...
-                if (webTask.IsCompleted && webTask.IsFaulted)
+                catch (Exception)
                 {
-                    var innerException = webTask.Exception?.InnerException;
-                    if (innerException != null)
-                        throw innerException;
+                    return;
                 }
+
+                await host.RunAsync(cancellationToken);
```
Comment (ganeshnj) [inline src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs:73]: It appears like a regression, previously when the host task (web app) exited, we were throwing an exception which is no longer true now.

I think `ServerModeSession` will kick in and re-start the server but we are losing the diagnostics here IMO.

Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/ServerModeCommand.cs:73]: Before we were using Task.WhenAny which causes an AggregateException to be thrown. We were catching that and throwing back the inner exception. Now we don't have that anymore and await host.RunAsync(cancellationToken) will just throw an exception same as before. The same behavior was retained.

**Review by ganeshnj — APPROVED** [review]

**Review by 96malhar — APPROVED** [review]

**Review by normj — APPROVED** [review]

**Pull request merged**